keps/sig-cluster-lifecycle/0035-20190128-addons-via-operators.md

```diff
@@ -0,0 +1,222 @@
+  (LocalDNS-Cache?) and figure out how it can be used despite being out-of-tree
+* Investigate use of webhooks to prevent accidental mutation of child objects
+* Investigate the RBAC story for addons - currently the operator must itself
+  have all the permissions that the addon needs, which is not really
+  least-privilege.  But it is not clear how to side-step this, nor that any of
+  the alternatives would be better or more secure.
+* Investigate use of patching mechanisms (as seen in `kubectl patch` and
+  `kustomize`) to support advanced tailoring of addons.  The goal here is to
+  make sure that everyone can use the addon operators, even if they “love it but
+  just need to change one thing”.  This ensures that the addon operators
+  themselves can remain bounded in scope and complexity.  Patching will fail if
+  the underlying addon changes dramatically, so we'll likely have a "patch
+  incompatible with new version" error - and generally addons should avoid
+  gratuitously changing their structure.
+* We should develop a convention (labels / owner-refs) so that we are able to
+  discover which CRDs are cluster-addons, and there is no confusion with
+  application operators.
+
+We expect the following functionality to be common to all operators for addons:
+
+* A CRD per addon
+* Common fields in spec that define the version and/or channel
+* Common fields in status that expose the current health & version information
+  of the addon
+* Addons follow a common structure, with the CR as root object, an Application
+  CR, consistent labels of all objects
+* Some form of protection or rapid reconciliation to prevent accidental
+  modification of child objects
+* Operators are declaratively driven, and can source manifests via https
+  (including mirrors), or from data stored in the cluster itself
+  (e.g. configmaps or cluster-bundle CRD, useful for airgapped)
+* Operators are able to expose different update behaviours: automatic immediate
+  updates; notification of update-available in status; purely manual updates
+* Operators are able to observe other CRs to perform basic sequencing
+* Addon manifests are able express an operator minimum version requirement, so
+  that an addon with new requirements can require that the operator be updated
+  first
+* Airgapped operation should be possible by combining a registry mirror and
+  storage of the underlying manifest in the cluster itself.
+
+
+An example can make this easier to understand, here is what a CRD instance for
+kube-proxy might look like:
+
+```yaml
+apiVersion: addons.sigs.k8s.io/v1alpha1
+kind: KubeProxy
+metadata:
+  name: default
+  namespace: kube-system
+spec:
+  clusterCidr: 100.64.0.0/10
+  version: 1.14.4
+status:
+  healthy: true
+```
+
+This particular manifest is pinned to `version: 1.14.4`.  We could also
+subscribe to a stream of updates with a field like `channel: stable`.
+
+### Risks and Mitigations
```

Comment:
I think test and verification for a release is also super important.

Reply:
Good point - added a paragraph specifically for that.  I think if we get to parity with the "embedded manifest" approach that we know and love, that'll help initially.  And I think we unlock approaches that are a little more flexible than embedding manifests, but I don't think we need to dictate anything there.